File: api/WebApp.Domain/Entities/Namespace.cs
using System.Diagnostics.CodeAnalysis;

namespace WebApp.Domain.Entities;

public sealed record Namespace
{
    public NamespaceId Id { get; init; }
    public NamespaceKind Kind { get; init; }
    public UserId? UserId { get; init; }
    public User? User { get; init; }
    public ICollection<Project> Projects { get; init; } = null!;

    [MemberNotNullWhen(true, nameof(UserId), nameof(User))]
    public bool IsUserNamespace => Kind == NamespaceKind.User;
}

File: www/src/lib/models/namespace.ts
import { isPlainObject } from 'is-what';
import type { Project } from './project';
import type { User } from './user';

export enum NamespaceKind {
	None = 'none',
	User = 'user',
}

export interface UserNamespace {
	id: string;
	kind: 'user';
	userId: string;
	user: User;
	projects: Project[];
}

export type Namespace = UserNamespace;

export function isUserNamespace(obj: unknown): obj is UserNamespace {
	return isPlainObject(obj) && 'kind' in obj && obj.kind === 'user';
}